Refactor curriculum schema to establish 1:1 relationship with person  issue  #100

fazzy12 opened the pull request:
- Update curriculum schema from version 1.0.0 to 2.0.0
- Modify schema to define a 1:1 relationship with a person, replacing the previous one-to-one relationship with a resource from the topics domain
- Remove person ID property and add topic ID to the resource object
- Update configuration file (config/curriculum.json) to reflect the schema version change
- Create meaningful test data in data/curriculum.json

fazzy12 committed:
Refactor curriculum schema to establish 1:1 relationship with person

FlatBallFlyer commented on src/mongosh/config/curriculum.json:
```diff
@@ -1,11 +1,11 @@
   "indexes": [
     {
-      "keys": {"personId": 1, "roadmapGroup": 1},
+      "keys": {"_id": 1, "roadmap": 1},
```
I think we can drop this index all together, it is no longer needed.

FlatBallFlyer commented on src/mongosh/schemas/curriculum-2.0.0.json:
```diff
@@ -0,0 +1,99 @@
+			"bsonType": "string",
+			"maxLength": 20
+		},
+		"resources": {
```
you will need to add a "name" property temporarily - just a type:string - we will remove this in a future iteration, so declare it but don't use it.

FlatBallFlyer commented on src/mongosh/schemas/curriculum-2.0.0.json:
```diff
@@ -0,0 +1,99 @@
+			"minimum": 0,
+			"maximum": 5
+		}
+	},
```
These properties - status -> rating are all part of the resources array.

FlatBallFlyer commented on src/mongosh/data/curriculum.json:
This file will need to pass the schema definitions.

fazzy12 replied:
tried to run the test, but kept falling.  would love to know what I am doing wrongly.

FlatBallFlyer reviewed (changes requested):
You are close, let's get these changes made and update the test data - when you think it is all ready you can do this to test it
```bash
cd src/mongosh
./test.sh
```
That should run the scripts to apply the schema and load the test data. If it is all setup correctly you will see a Completed Successfully message. If you get errors, then you need to fix your configurations.

fazzy12 committed:
Fix index and schema issues in curriculum configuration

fazzy12 committed:
Fix index and schema issues in curriculum configuration

fazzy12 commented:
- Removed unnecessary index on personId and roadmapGroup in curriculum configuration
- Added temporary "name" property in curriculum schema as per request for future iteration
- Ensured that status to rating properties are included in the resources array in the curriculum schema
- Updated curriculum data file to pass the schema definitions

FlatBallFlyer commented on src/mongosh/schemas/curriculum-2.0.0.json:
```diff
@@ -0,0 +1,103 @@
+				"required": ["type", "review", "roadmap"],
+				"properties": {
+					"type": {
+						"description": "Type of the resource (topic or adhoc)",
```
I think this is your first problem, type does not have a "bsonType" property which makes this schema invalid.

FlatBallFlyer commented on src/mongosh/schemas/curriculum-2.0.0.json:
```diff
@@ -0,0 +1,103 @@
+						"bsonType": "string",
+						"maxLength": 256
+					},
+					"roadmap": {
```
Same problem here, missing a bsontype indicator

FlatBallFlyer commented on src/mongosh/schemas/curriculum-2.0.0.json:
```diff
@@ -0,0 +1,103 @@
+					}
+				},
+				"oneOf": [
+					{
```
I like what you're documenting here, but I've not used the oneOf constraint before in a mongosh schema, and I don't think that this structure has any impact unless properties are required. We should not have this many required properties because of the auto save functionality that we're using. You should think about the data as a new document is created, one property at a time. I think the only required field we should have this entire document is _ID

FlatBallFlyer reviewed (changes requested):
I made some notes that I think outline where you're having problems with your scheme of validation. The hint was in the error message when I ran the test,
```
TypeError: Cannot read properties of undefined (reading 'bsonType')
```

fazzy12 committed:
refactor: Update schema validation for curriculum document

fazzy12 commented:
- Removed unnecessary "oneOf" constraint in the schema validation for the curriculum document.
- Revised the list of required properties. Now, only the "_id" property is marked as required, reflecting the auto-save functionality used in the system.
- Made minor adjustments to property descriptions for clarity and consistency.

fazzy12 commented:
- Removed unnecessary "oneOf" constraint in the schema validation for the curriculum document.
- Revised the list of required properties. Now, only the "_id" property is marked as required, reflecting the auto-save functionality used in the system.
- Made minor adjustments to property descriptions for clarity and consistency.

Commit:
Removed system enumerators for curriculum

fazzy12 committed:
refactor: Update schema validation for curriculum document

fazzy12 committed:
Merge remote-tracking branch 'origin/refactor/curriculum-1-to-1-to-person' into refactor/curriculum-1-to-1-to-person
